### PR TITLE
[BEAM-11659] Extract common payload serialization behavior for Row transforms.

### PR DESCRIPTION
**Pull request by dpcollins-google**
In the process, enable using proto and thrift serialization with Pub/Sub.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by dpcollins-google** [comment]
R: @amaliujia

**Comment by dpcollins-google** [comment]
R: @robinyqiu

**Comment by amaliujia** [comment]
This seems a bit large refactoring so I will need more time to review. Meanwhile, if there is a way to split this change to smaller changes that will be very helpful.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/payloads/AvroPayloadSerializerProvider.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+@Experimental(Kind.SCHEMAS)
+@AutoService(PayloadSerializerProvider.class)
+public class AvroPayloadSerializerProvider implements PayloadSerializerProvider {
+  @Override
+  public String identifier() {
+    return "avro";
+  }
+
+  @Override
+  public PayloadSerializer getSerializer(Schema schema, Map<String, Object> tableParams) {
```
Comment (amaliujia) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/payloads/AvroPayloadSerializerProvider.java:38]: Question: Is `Map<String, Object> tableParams` the place for pubsub/kafka attributes?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProvider.java [hunk 2]
```diff
@@ -69,65 +58,16 @@ public BeamSqlTable buildBeamSqlTable(Table table) {
...
-        throw new IllegalStateException(
-            "Could not instantiate the thrift protocol factory class", e);
-      }
-      return new BeamKafkaThriftTable<>(
-          schema, bootstrapServers, topics, thriftClass, thriftProtocolFactory);
-    } catch (ClassNotFoundException e) {
-      throw new IllegalArgumentException("Incorrect thrift class provided: " + thriftClassName);
+    String payloadFormat =
+        properties.containsKey("format") ? properties.getString("format") : "csv";
+    if (payloadFormat.equals("csv")) {
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProvider.java:63]: Can you add a comment here to describe why CSV is a special why that does not go through PayloadSerializerKafkaTable?

**Comment by amaliujia** [comment]
Seems there is checkstyle issue. 

You try to run `./gradlew spotlessApply`. If this command cannot fix that, you will have to check the report to manually fix issues.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProvider.java (hunk 2 above)
Comment (dpcollins-google) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/KafkaTableProvider.java:69]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/payloads/AvroPayloadSerializerProvider.java (hunk 1 above)
Comment (dpcollins-google) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/payloads/AvroPayloadSerializerProvider.java:38]: Its a place for Proto/Thrift parameters. Added a comment.

File: gradle/wrapper/gradle-wrapper.properties [hunk 3]
```diff
@@ -1,5 +1,5 @@
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.8-all.zip
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.8-bin.zip
-zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
+zipStoreBase=GRADLE_USER_HOME
```
[the remark this answers is not in the document]
Comment (dpcollins-google) [reply gradle/wrapper/gradle-wrapper.properties]: Done.

**Comment by dpcollins-google** [comment]
> Seems there is checkstyle issue.
> 
> You try to run `./gradlew spotlessApply`. If this command cannot fix that, you will have to check the report to manually fix issues.

Fixed.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/payloads/AvroPayloadSerializerProvider.java (hunk 1 above)
Comment (dpcollins-google) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/io/payloads/AvroPayloadSerializerProvider.java:38]: To clarify: Added a comment to the interface.

**Comment by amaliujia** [comment]
I will give another pass soon.

cc @robinyqiu could you also check on this PR as this is a relatively large refactoring for table providers.

**Comment by robinyqiu** [comment]
Yes I am looking at it. I took a look at it last week but did not finish the review. It will take some time since this is a big change.

**Comment by robinyqiu** [comment]
Run Java PreCommit

**Review by robinyqiu — APPROVED** [review]
LGTM (after the precommit tests pass)

**Comment by dpcollins-google** [comment]
Run Java PreCommit

**Comment by dpcollins-google** [comment]
Run Java PreCommit

**Comment by dpcollins-google** [comment]
Run Java_Examples_Dataflow PreCommit

**Review by amaliujia — APPROVED** [review]

**Comment by amaliujia** [comment]
run SQL postcommit

**Comment by amaliujia** [comment]
@dpcollins-google 
Can you make all CI check pass so I can merge this PR?

**Comment by dpcollins-google** [comment]
Run Java PreCommit

**Comment by dpcollins-google** [comment]
Run Java PreCommi

**Comment by dpcollins-google** [comment]
Run Java PreCommit

**Comment by dpcollins-google** [comment]
Run Java PreCommit

**Comment by TheNeuralBit** [comment]
CC: @ccciudatu FYI this is the kind of interface we had talked about making to generalize the serialization for use in PubSub and Kafka (and other message queues)

Thanks @dpcollins-google

**Comment by TheNeuralBit** [comment]
We should update https://beam.apache.org/documentation/dsls/sql/extensions/create-external-table/#pubsub with the new supported formats, and maybe CHANGES.md too

**Pull request merged**